File: clients/rust/src/wire.rs
//! Wire-protocol types — mirrors pkg/protocol/wire.go exactly.
//!
//! All field names and JSON tags match Go 1:1. `omitempty` in Go →
//! `#[serde(skip_serializing_if = "...")]` here. Timestamps are RFC3339 UTC
//! via the `time` crate's well-known Rfc3339 format.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::encrypt::{EncryptedPayload, SealedPayload};
use crate::sign::SignedEnvelope;

pub const VERSION: &str = "1";

/// Module types the Rust agent declares support for. v1 (JSON) module
/// types were removed from the server; only v2 (INI) remains. Older
/// agents whose appliers still parse JSON must be upgraded — the
/// dispatcher silently skips unknown module types so the agent won't
/// crash, but it also won't apply anything.
pub const CAPABILITIES: &[&str] = &[
    "wifi.v2",
    "ssh.authorized_keys.v2",
    "c8y.v2",
    "files.v2",
    "hook.v2",
    "passwd.v2",
];

// ---- EnrollRequest ----------------------------------------------------------

/// Sent by the device to request a provisioning bundle.
/// Signed with the device's long-lived Ed25519 identity key.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EnrollRequest {
    pub protocol_version: String,
    pub nonce: String, // base64(random 16 bytes)

    #[serde(with = "rfc3339_z")]
    pub timestamp: DateTime<Utc>,

    pub device_id: String,
    pub public_key: String, // base64(Ed25519 pub, 32 bytes)

    #[serde(skip_serializing_if = "Option::is_none")]
    pub ephemeral_x25519: Option<String>, // base64(X25519 pub, 32 bytes)

    #[serde(skip_serializing_if = "is_false")]
    pub encrypt_bundle: bool,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub bootstrap_token: Option<String>,

    pub facts: DeviceFacts,

    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub capabilities: Vec<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<std::collections::HashMap<String, String>>,
}

/// Device identity facts included in every enroll request.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct DeviceFacts {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub machine_id: Option<String>,

    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub mac_addresses: Vec<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub serial: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub hostname: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub os: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub arch: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_version: Option<String>,
}

// ---- EnrollResponse ---------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum EnrollStatus {
    Accepted,
    Pending,
    Rejected,
}

/// Server's reply to an enroll request.
#[derive(Debug, Clone, Deserialize)]
pub struct EnrollResponse {
    pub protocol_version: String,
    pub status: EnrollStatus,

    #[serde(default)]
    pub reason: Option<String>,

    #[serde(default)]
    pub retry_after: Option<u32>, // seconds

    #[serde(default)]
    pub bundle: Option<SignedEnvelope>,

    #[serde(default)]
    pub encrypted_bundle: Option<EncryptedPayload>,

    /// Server's UTC clock at response time. Devices whose clocks are not yet
    /// synced can compute a correction offset from this field and retry.
    #[serde(default, skip_serializing_if = "Option::is_none", with = "opt_rfc3339_z")]
    pub server_time: Option<DateTime<Utc>>,
}

// ---- ProvisioningBundle -----------------------------------------------------

/// The configuration delivered to a device after enrolment.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProvisioningBundle {
    pub protocol_version: String,
    pub device_id: String,

    #[serde(with = "rfc3339_z")]
    pub issued_at: DateTime<Utc>,

    #[serde(default, skip_serializing_if = "Option::is_none", with = "opt_rfc3339_z")]
    pub expires_at: Option<DateTime<Utc>>,

    pub modules: Vec<Module>,
}

/// A single configuration module inside a bundle.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Module {
    #[serde(rename = "type")]
    pub module_type: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub payload: Option<serde_json::Map<String, serde_json::Value>>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub sealed: Option<SealedPayload>,

    /// `[]byte` in Go → base64 string on the wire.
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        serialize_with = "serialize_optional_base64",
        deserialize_with = "deserialize_optional_base64"
    )]
    pub raw_payload: Option<Vec<u8>>,
}

// ---- Acknowledgement ---------------------------------------------------------

/// Sent by the device after applying a bundle (optional in current protocol).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Acknowledgement {
    pub protocol_version: String,
    pub device_id: String,

    #[serde(with = "rfc3339_z")]
    pub bundle_issued_at: DateTime<Utc>,

    pub results: Vec<ModuleResult>,
}

/// Per-module outcome.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModuleResult {
    #[serde(rename = "type")]
    pub module_type: String,
    pub ok: bool,

    #[serde(skip_serializing_if = "is_false")]
    pub skipped: bool,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub output: Option<String>,
}

// ---- serde helpers ----------------------------------------------------------

fn is_false(b: &bool) -> bool {
    !b
}

use base64::{engine::general_purpose::STANDARD, Engine as _};

/// Serialize/deserialize DateTime<Utc> as RFC3339 with 'Z' suffix, seconds precision.
/// Matches Go's time.MarshalJSON format for timestamps without nanoseconds.
mod rfc3339_z {
    use chrono::{DateTime, SecondsFormat, Utc};
    use serde::{Deserializer, Serializer};

    pub fn serialize<S: Serializer>(dt: &DateTime<Utc>, s: S) -> Result<S::Ok, S::Error> {
        s.serialize_str(&dt.to_rfc3339_opts(SecondsFormat::Secs, true))
    }

    pub fn deserialize<'de, D: Deserializer<'de>>(d: D) -> Result<DateTime<Utc>, D::Error> {
        let s = String::deserialize(d)?;
        s.parse::<DateTime<Utc>>().map_err(serde::de::Error::custom)
    }

    use serde::Deserialize;
}

mod opt_rfc3339_z {
    use chrono::{DateTime, SecondsFormat, Utc};
    use serde::{Deserialize, Deserializer, Serializer};

    pub fn serialize<S: Serializer>(dt: &Option<DateTime<Utc>>, s: S) -> Result<S::Ok, S::Error> {
        match dt {
            Some(dt) => s.serialize_str(&dt.to_rfc3339_opts(SecondsFormat::Secs, true)),
            None => s.serialize_none(),
        }
    }

    pub fn deserialize<'de, D: Deserializer<'de>>(
        d: D,
    ) -> Result<Option<DateTime<Utc>>, D::Error> {
        let opt: Option<String> = Option::deserialize(d)?;
        match opt {
            Some(s) => s
                .parse::<DateTime<Utc>>()
                .map(Some)
                .map_err(serde::de::Error::custom),
            None => Ok(None),
        }
    }
}
use serde::{Deserializer, Serializer};

fn serialize_optional_base64<S: Serializer>(
    v: &Option<Vec<u8>>,
    s: S,
) -> Result<S::Ok, S::Error> {
    match v {
        Some(bytes) => s.serialize_str(&STANDARD.encode(bytes)),
        None => s.serialize_none(),
    }
}

fn deserialize_optional_base64<'de, D: Deserializer<'de>>(
    d: D,
) -> Result<Option<Vec<u8>>, D::Error> {
    let opt: Option<String> = Option::deserialize(d)?;
    match opt {
        Some(s) => STANDARD.decode(&s).map(Some).map_err(serde::de::Error::custom),
        None => Ok(None),
    }
}

File: pkg/protocol/wire.go
// Package protocol defines the wire format shared between the ZTP server and
// device agents. The format is transport-agnostic and works equally well over
// HTTPS, mDNS-discovered LAN, or BLE GATT relays.
//
// All payloads are JSON. Signatures are detached Ed25519 over a canonicalised
// (RFC 8785 / JCS) representation of the payload object so the Go and shell
// agents produce identical signing inputs.
package protocol

import "time"

// Version is the current wire protocol version. The server and agent both
// emit this value and reject mismatched majors.
const Version = "1"

// EnrollRequest is sent by a device to claim an identity and request a
// provisioning bundle. It is signed by the device's long-lived identity key
// (Ed25519). The PublicKey field carries the verifier needed by the server,
// allowing trust-on-first-use flows.
type EnrollRequest struct {
	ProtocolVersion string    `json:"protocol_version"`
	Nonce           string    `json:"nonce"`     // base64(random 16 bytes)
	Timestamp       time.Time `json:"timestamp"` // RFC3339, UTC
	DeviceID        string    `json:"device_id"`
	PublicKey       string    `json:"public_key"`                 // base64 Ed25519 pub
	EphemeralX25519 string    `json:"ephemeral_x25519,omitempty"` // base64; for end-to-end encryption (whole bundle and/or sensitive modules)
	// EncryptBundle requests that the entire SignedEnvelope be wrapped in an
	// EncryptedPayload addressed to EphemeralX25519. Useful when the
	// transport itself is untrusted (e.g. a BLE relay). Independent of
	// per-module sealing, which the server applies whenever it has a
	// sensitive payload AND EphemeralX25519 is present.
	EncryptBundle  bool              `json:"encrypt_bundle,omitempty"`
	BootstrapToken string            `json:"bootstrap_token,omitempty"`
	Facts          DeviceFacts       `json:"facts"`
	Capabilities   []string          `json:"capabilities,omitempty"` // module types the device supports
	Metadata       map[string]string `json:"metadata,omitempty"`
}

// DeviceFacts is the operator-visible identity surface of a device. It is what
// shows up in the "pending approval" list of the web UI.
type DeviceFacts struct {
	MachineID    string   `json:"machine_id,omitempty"`
	MACAddresses []string `json:"mac_addresses,omitempty"`
	Serial       string   `json:"serial,omitempty"`
	Model        string   `json:"model,omitempty"`
	Hostname     string   `json:"hostname,omitempty"`
	OS           string   `json:"os,omitempty"`
	Arch         string   `json:"arch,omitempty"`
	// OSPrettyName is the human-readable OS name from /etc/os-release
	// (PRETTY_NAME field), e.g. "Debian GNU/Linux 12 (bookworm)".
	OSPrettyName string `json:"os_pretty_name,omitempty"`
	AgentVersion string `json:"agent_version,omitempty"`
}

// SignedEnvelope wraps a JSON document with a detached signature.
// Payload is the canonicalised JSON bytes (base64 encoded) so verifiers do not
// have to re-canonicalise. KeyID identifies which key signed it (e.g. "device"
// or a server key fingerprint).
type SignedEnvelope struct {
	ProtocolVersion string `json:"protocol_version"`
	KeyID           string `json:"key_id"`
	Algorithm       string `json:"alg"`       // "ed25519"
	Payload         string `json:"payload"`   // base64(canonical JSON)
	Signature       string `json:"signature"` // base64(sig)
}

// EnrollStatus is the high-level result returned to a device.
type EnrollStatus string

const (
	StatusAccepted EnrollStatus = "accepted"
	StatusPending  EnrollStatus = "pending"
	StatusRejected EnrollStatus = "rejected"
)

// EnrollResponse is the server's reply. When Status == StatusAccepted the
// Bundle field contains a signed (and optionally encrypted) ProvisioningBundle.
type EnrollResponse struct {
	ProtocolVersion string          `json:"protocol_version"`
	Status          EnrollStatus    `json:"status"`
	Reason          string          `json:"reason,omitempty"`
	RetryAfter      int             `json:"retry_after,omitempty"` // seconds; for "pending"
	Bundle          *SignedEnvelope `json:"bundle,omitempty"`
	// EncryptedBundle, if present, is an X25519+ChaCha20-Poly1305 ciphertext
	// over the SignedEnvelope JSON. Used when transport is untrusted (BLE relay).
	EncryptedBundle *EncryptedPayload `json:"encrypted_bundle,omitempty"`
	// TextManifest, if present, is an alternative bundle representation as
	// signed line-based "key=value" text. It is omitted from JSON responses
	// (it would just duplicate Bundle) and rendered only when the client
	// asks for `Accept: text/plain`.
	TextManifest *SignedEnvelope `json:"-"`
	// ServerTime is the server's UTC clock at the time of the response.
	// Devices whose clocks are not yet synced (e.g. before NTP via a
	// newly-provisioned network) can compute a correction offset from this
	// field and retry with an adjusted timestamp.
	ServerTime *time.Time `json:"server_time,omitempty"`
}

// EncryptedPayload carries an end-to-end-encrypted payload addressed to the
// device's ephemeral X25519 key from EnrollRequest.EphemeralX25519.
type EncryptedPayload struct {
	Algorithm  string `json:"alg"`        // "x25519-chacha20poly1305"
	ServerKey  string `json:"server_key"` // base64 X25519 pub
	Nonce      string `json:"nonce"`      // base64 12 bytes
	Ciphertext string `json:"ciphertext"` // base64
}

// ProvisioningBundle is the configuration delivered to a device. Modules are
// processed independently; unknown module types must be skipped, not fail the
// whole provisioning.
type ProvisioningBundle struct {
	ProtocolVersion string    `json:"protocol_version"`
	DeviceID        string    `json:"device_id"`
	IssuedAt        time.Time `json:"issued_at"`
	ExpiresAt       time.Time `json:"expires_at,omitempty"`
	Modules         []Module  `json:"modules"`
}

// Module is one piece of configuration. Type uniquely identifies a handler
// (e.g. "wifi.v2", "ssh.authorized_keys.v2"). Payload is opaque JSON whose
// schema is defined by the handler.
//
// RawPayload, if non-nil, is used verbatim as the module's payload bytes
// (still base64-encoded into the manifest line). It exists to support
// non-JSON payload formats — e.g. wifi.v2 ships an INI document so the
// applier needs no JSON parser. When RawPayload is set, Payload is ignored
// and need not be valid JSON-shaped data.
//
// Sealed, if non-nil, replaces Payload/RawPayload with an end-to-end-encrypted
// ciphertext addressed to the device's ephemeral X25519 key. The server
// populates this for modules that carry secrets the server itself should not
// retain in plaintext logs/audit/SQLite (e.g. a Cumulocity enrollment token).
// The agent decrypts before dispatching to the applier; everything downstream
// of the dispatcher behaves as if the module had been delivered in the clear.
//
// Sensitive is a server-side hint, not part of the wire format. Providers set
// it on modules that contain secrets so the engine knows to seal them before
// signing the bundle. It is intentionally json-omitted because once the engine
// has sealed the payload there is nothing for the wire-format reader to do
// with this flag.
type Module struct {
	Type       string         `json:"type"`
	Payload    map[string]any `json:"payload,omitempty"`
	Sealed     *SealedPayload `json:"sealed,omitempty"`
	RawPayload []byte         `json:"raw_payload,omitempty"`
	Sensitive  bool           `json:"-"`
}

// SealedPayload is a per-module ciphertext addressed to the device's ephemeral
// X25519 key (EnrollRequest.EphemeralX25519). It uses the same primitive as
// EncryptedPayload (X25519 + ChaCha20-Poly1305) but lives inside a Module so a
// single bundle can mix sealed and clear modules. Format describes how the
// agent should treat the decrypted bytes: "json" → unmarshal into the
// module's Payload map; "raw" → hand directly to the applier as RawPayload.
type SealedPayload struct {
	Algorithm    string `json:"alg"`           // "x25519-chacha20poly1305"
	EphemeralPub string `json:"ephemeral_pub"` // base64 X25519 pub the agent uses for ECDH
	Nonce        string `json:"nonce"`         // base64 12 bytes
	Ciphertext   string `json:"ciphertext"`    // base64
	Format       string `json:"format"`        // "json" | "raw"
}

// Acknowledgement is sent by the device after applying (or failing to apply)
// a bundle, to close the loop with the server.
type Acknowledgement struct {
	ProtocolVersion string         `json:"protocol_version"`
	DeviceID        string         `json:"device_id"`
	BundleIssuedAt  time.Time      `json:"bundle_issued_at"`
	Results         []ModuleResult `json:"results"`
}

// ModuleResult is the per-module outcome of applying a bundle.
type ModuleResult struct {
	Type    string `json:"type"`
	OK      bool   `json:"ok"`
	Skipped bool   `json:"skipped,omitempty"` // true when no applier handled the type
	Error   string `json:"error,omitempty"`
	Output  string `json:"output,omitempty"`
}
